### PR TITLE
added supernova class

### PR DESCRIPTION
**Pull request by jpierel14**
initial class for supernovae, pending discussion on exactly what needs to go in here

### CODE REVIEW

File: slsim/Sources/supernovae.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import numpy as np
+import numpy.random as random
+import sncosmo
+from slsim.Util import param_util
+
+
+class Supernovae(sncosmo.Model):
+    """Class describing elliptical galaxies."""
+
```
Comment (nkhadka21) [inline slsim/Sources/supernovae.py:9]: Class describing supernovae.

File: slsim/Sources/supernovae.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        :type redshift: float
+        :param sn_type: Supernova type (Ia, Ib, Ic, IIP, etc.)
+        :type sn_type: str
+        :param absolute_mag: Absolute magnitude of the supernova
+        :type absolute_mag: float
+        :param absolute_mag_band: Band used to normalize to absolute magnitude
+        :type absolute_mag_band: str or `~sncosmo.Bandpass`
+        :param absolute_mag_zpsys: AB or Vega
+        :type absolute_mag_zpsys: str
+        """
```
Comment (nkhadka21) [inline slsim/Sources/supernovae.py:35]: Can you develop this class as a class which takes list of simulated supernovae (astropy table) and manages it as you want. You can write a code to simulate supernovae population outside of this class and within the Source directory. Does it make sense?

**Review by nkhadka21 — CHANGES_REQUESTED** [review]
Hi @jpierel14 , thanks alot for this PR! I have few comments. If my comments are not clear, please let me know. Also, it would be nice if you could write test functions for your code.

File: slsim/Sources/supernovae.py (hunk 2 above)
Comment (nkhadka21) [reply slsim/Sources/supernovae.py:35]: Hi @jpierel14, I think you can develop Supernovae class to generate supernovae catalog in the form given in /slsim/Sources/SupernovaeData/supernovae_data.pkl. Then, we can call that class inside Lens class whenever supernovae catalog is needed. Thank you.

**Comment by jpierel14** [comment]
@sibirrer @nkhadka21 New commit has a version of the changes we discussed today, let me know what you think

**Comment by sibirrer** [comment]
Thank you @jpierel14 ! The pre-commit failed. This one is hopefully easy to fix. Then for the CI, the import of sncosmo failed. Can you add sncosmo (plus version) on the requirements.txt file?

File: slsim/Sources/supernovae.py (hunk 2 above)
Comment (jpierel14) [reply slsim/Sources/supernovae.py:65]: @nkhadka21 sorry this conversation is a bit out of date now but is there still some sort of other interface you'd be interested in here?

**Comment by jpierel14** [comment]
Okay @sibirrer better now, these other two are just because I don't have tests for every function?

File: slsim/Sources/supernovae.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            )
+        else:
+            print(
+                "Warning, you should use self.set_source_peakabsmag or sefl.set_source_peakmag to set the amplitude."
+            )
+
+    def get_apparent_magnitude(self, time, band, zpsys="AB"):
+        """Function to return apparent magnitude of a SN for a given band and time.
+
+        :param time: The observer-frame time array to evaluate the model
```
Comment (nkhadka21) [inline slsim/Sources/supernovae.py:94]: What is the unit of time? It would be helpful to specify unit.

File: slsim/Sources/supernovae.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        :type time: `~np.ndarray` or list
+        :param band: The bandpass to evaluate the model over
+        :type band: str or `~sncosmo.Bandpass`
+        :param zpsys: Optional, AB or Vega (AB default)
+        :type zpsys: str
```
Comment (nkhadka21) [inline slsim/Sources/supernovae.py:99]: :return: ?, need a docstring for return.

File: slsim/Sources/supernovae.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        if self._sn_type == "Ia":
+            self.set(**{"c": np.random.normal(0, 0.1), "x1": np.random.normal(0, 1)})
+
+    def set_random_sed_model(self, sn_type, random_seed=42):
+        """Function to set a random SED model for a given SN type.
+
+        :param sn_type: Supernova type (Ia, Ib, Ic, IIP, etc.)
+        :type sn_type: str
+        :param random_seed: Random seed for randomization
+        :type random_seed: int
```
Comment (nkhadka21) [inline slsim/Sources/supernovae.py:178]: :return: ?, need a docstring for return.

File: slsim/Sources/supernovae.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        self, sn_type, absolute_mag_distribution=None, random_seed=42
+    ):
+        """Function to get a reasonable absolute mag for a given SN type.
+
+        :param sn_type: Supernova type (Ia, Ib, Ic, IIP, etc.)
+        :type sn_type: str
+        :param absolute_mag_distribution: A function that returns an absolute mag
+        :type absolute_mag_distribution: func
+        :param random_seed: Random seed for randomization
+        :type random_seed: int
```
Comment (nkhadka21) [inline slsim/Sources/supernovae.py:220]: :return: ?, need a docstring for return.

**Review by nkhadka21 — CHANGES_REQUESTED** [review]
Hi @jpierel14, thank you for the update! I have few minor comments. I will look at this carefully soon.

File: slsim/Sources/supernovae.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        :param band: The bandpass to evaluate the model over
+        :type band: str or `~sncosmo.Bandpass`
+        :param zpsys: Optional, AB or Vega (AB default)
+        :type zpsys: str
+        """
+
+        return self.bandmag(band, zpsys, time)
+
+
+class RandomizedSupernova(Supernova):
```
Comment (nkhadka21) [inline slsim/Sources/supernovae.py:105]: It would be good to have each class in a separate file with an associated file name.

**Review by nkhadka21 — CHANGES_REQUESTED** [review]

File: slsim/Sources/supernovae.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        :type redshift: float
+        :param sn_type: Supernova type (Ia, Ib, Ic, IIP, etc.)
+        :type sn_type: str
+        :param absolute_mag: Absolute magnitude of the supernova
+        :type absolute_mag: float
+        :param absolute_mag_band: Band used to normalize to absolute magnitude
+        :type absolute_mag_band: str or `~sncosmo.Bandpass`
+        :param peak_apparent_mag: Peak apparent mag of the supernova
+        :type peak_apparent_mag: str or `~sncosmo.Bandpass`
+        :param peak_apparent_mag_band: Band used to normalize to apparent magnitude
```
Comment (nkhadka21) [inline slsim/Sources/supernovae.py:58]: Can we pass LSST filter to this class so that we get apparent magnitude in LSST fiters?

**Review by nkhadka21 — CHANGES_REQUESTED** [review]

File: slsim/Sources/supernovae.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            absolute_mag = np.random.normal(mu, sigma)
+            self._absolute_mag_band = "bessellb"
+        else:
+            absolute_mag = absolute_mag_distribution()
+        return absolute_mag
+
+
+if __name__ == "__main__":
+
+    test = RandomizedSupernova("IIP")
```
Comment (nkhadka21) [inline slsim/Sources/supernovae.py:236]: Do we need these last two line here?

**Review by nkhadka21 — CHANGES_REQUESTED** [review]

File: slsim/Sources/supernovae.py (hunk 3 above)
Comment (jpierel14) [reply slsim/Sources/supernovae.py:94]: done

**Comment by jpierel14** [comment]
Okay @nkhadka21 , try that, or should I resolve these codecov errors too?

**Comment by nkhadka21** [comment]
> Okay @nkhadka21 , try that, or should I resolve these codecov errors too?

Thanks @jpierel14! It would be great if you could resolve codcov error.

**Comment by nkhadka21** [comment]
Hi @jpierel14 , I merged this PR. Thank you.

**Pull request merged**